File: apps/server/internal/common/http.go
package common

type HTTPResponse struct {
	Result interface{} `json:"result,omitempty"`

	Code    string `json:"code,omitempty"`
	Message string `json:"message,omitempty"`
}

type PaginatedResult[T interface{}] struct {
	Count int `json:"count"`
	Items []T `json:"items"`

	Offset int `json:"offset"`
	Limit  int `json:"limit"`
}

var EmptyHTTPResponse = HTTPResponse{
	Result: struct{}{},
}

File: packages/core/common/http.ts
import { z } from "zod";

export const HTTPSuccessResponseSchema = <T extends z.ZodType>(schema: T) =>
	z.object({
		result: schema,
	});

export const HTTPErrorResponseSchema = <
	codes extends [string, ...string[]],
	T extends z.ZodEnum<codes>,
>(
	codes?: T
) =>
	z.object({
		code: codes ? codes : z.string(),
		message: z.string(),
	});

export const PaginatedResultSchema = <T extends z.ZodType>(schema: T) =>
	z.object({
		count: z.number(),
		items: z.array(schema),
		offset: z.number(),
		limit: z.number(),
	});
